### PR TITLE
Refactoring/gcp/gce

### PR DESCRIPTION
**Pull request by Remi05**
Migrated Google Compute Engine to the new architecture. Also adds support for filtering GCE Instances by Zone and GCE Subnetworks by Region.

### CODE REVIEW

File: ScoutSuite/providers/gcp/resources/gce/firewalls.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        firewall_dict['network'] = raw_firewall['network'].split('/')[-1]
+        firewall_dict['network_url'] = raw_firewall['network']
+        firewall_dict['priority'] = raw_firewall['priority']
+        firewall_dict['source_ranges'] = raw_firewall.get('sourceRanges', [])
+        firewall_dict['source_tags'] = raw_firewall.get('sourceTags', [])
+        firewall_dict['target_tags'] = raw_firewall.get('targetTags', [])
+        firewall_dict['direction'] = raw_firewall['direction']
+        firewall_dict['disabled'] = raw_firewall['disabled']
+
+        # Parse FW rules
```
Comment (Remi05) [inline ScoutSuite/providers/gcp/resources/gce/firewalls.py]: Clean this up / move it to a separate method.

File: ScoutSuite/providers/gcp/resources/gce/instances.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        return description if description else 'N/A'
+
+    def _is_block_project_ssh_keys_enabled(self, raw_instance):
+        metadata = self._metadata_to_dict(raw_instance['metadata'])
+        return metadata.get('block-project-ssh-keys') == 'true'
+
+    def _metadata_to_dict(self, metadata):
+        return dict((item['key'], item['value']) for item in metadata['items']) if 'items' in metadata else {}
+
+    async def _get_common_instance_metadata_dict(self):
```
Comment (misg) [inline ScoutSuite/providers/gcp/resources/gce/instances.py]: This method should be moved to GCE facade and those metadata should get attached to their instance objects when calling `get_instances` in order to maintain the separation of concerns principle (between the fetching and parsing parts) and also avoid performance issues (in this case, since the parsing is `async` because it does some fetching, the `for loop` in `fetch_all` is not really efficient and could lead to poor performances).
Although, regarding our schedule, I don't mind (do you @Aboisier ?) if you simply add a `TODO` to move forward and merge those PRs so we can unify a last time our refactoring.

File: ScoutSuite/providers/gcp/resources/gce/service.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        # and Zones classes, but having a "resource_name_count" field in their 
+        # dictionary causes errors in the rule engine since they are accessed
+        instances_count = 0
+        subnetworks_count = 0
+        for project in self['projects'].values():
+            for zone in project['zones'].values():
+                instances_count += zone['instances_count']
+            for region in project['regions'].values():
+                subnetworks_count += region['subnetworks_count']
+        self['instances_count'] = instances_count
```
Comment (misg) [inline ScoutSuite/providers/gcp/resources/gce/service.py]: I think a one liner would be more pythonic (don't forget to remove lines 24 to 30 if you agree with that):
```suggestion
        self['instances_count'] = sum([zone['instances_count'] for zone in project['zones'].values() for project in self['projects'].values()])
```

File: ScoutSuite/providers/gcp/resources/gce/service.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        self['subnetworks_count'] = subnetworks_count
```
Comment (misg) [inline ScoutSuite/providers/gcp/resources/gce/service.py]: Same here:
```suggestion
        self['subnetworks_count'] = sum([region['subnetworks_count'] for region in project['regions'].values() for project in self['projects'].values()])
```

**Comment by Aboisier** [comment]
@remi05 is this review-ready?

File: ScoutSuite/providers/gcp/provider.py [hunk 5]
```diff
@@ -87,7 +87,7 @@ def preprocessing(self, ip_ranges=None, ip_ranges_name_key=None):
         """
         ip_ranges = [] if ip_ranges is None else ip_ranges
 
-        self._match_instances_and_snapshots()
+        # self._match_instances_and_snapshots()
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/provider.py]: What's up with that?

File: ScoutSuite/providers/gcp/resources/gce/instances.py (hunk 2 above)
Comment (Aboisier) [reply ScoutSuite/providers/gcp/resources/gce/instances.py]: It would be ideal is if was fixed before merging, but I understand we're running out of time now. :/

File: ScoutSuite/output/data/html/partials/gcp/left_menu_for_gcp_region.html [hunk 6]
```diff
@@ -0,0 +1,38 @@
+<!-- Generic left-menu link partial for region-then-resource -->
+<script id="left_menu_for_gcp_region.partial" type="text/x-handlebars-template">
+  <div id="services.{{service_name}}.projects.id.regions.id.{{resource_type}}.list">
+  <div class="list-group" >
+    <div class="list-group-item">
+      <a href="javascript:showRowWithItems('services.{{service_name}}.projects.id.regions.id.{{resource_type}}')">Show all <span class="badge float-right btn-info">{{get_value_at 'metadata' service_group service_name 'resources' resource_type 'count'}}</span></a>
```
Comment (Aboisier) [inline ScoutSuite/output/data/html/partials/gcp/left_menu_for_gcp_region.html]: `get_value_at` was renamed, it's now `getValueAt` I believe.

**Comment by Aboisier** [comment]
> I'm also wondering if the `Regions` and `Zones` classes you defined couldn't be moved "up" (to `providers/gcp/resources/`, at the same level than `GCPCompositeResources`) in order to be "abstract" (like AWS `Regions` and `VPCs`) and be used for potential other GCP services. What do you think @Aboisier ?

Yup, looks like that would make sense! Ì haven't worked with GCP enough to know if there are other services which use regions. I'm guessing there are.

File: ScoutSuite/providers/gcp/resources/gce/service.py (hunk 3 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/resources/gce/service.py]: Sure, I was mainly looking to avoid super long lines tbh

**Comment by Remi05** [comment]
> > I'm also wondering if the `Regions` and `Zones` classes you defined couldn't be moved "up" (to `providers/gcp/resources/`, at the same level than `GCPCompositeResources`) in order to be "abstract" (like AWS `Regions` and `VPCs`) and be used for potential other GCP services. What do you think @Aboisier ?
> 
> Yup, looks like that would make sense! Ì haven't worked with GCP enough to know if there are other services which use regions. I'm guessing there are.

I tried to do that at first but it caused a lot of issues, it's a bit different than AWS' VPC and Region. Anyway, the Regions and Zones are currently only used in GCE's API (to fetch regions and zones, we use the Compute Engine client).

File: ScoutSuite/output/data/inc-scoutsuite/scoutsuite.js [hunk 7]
```diff
@@ -97,6 +98,12 @@ function onPageLoad() {
 
 };
 
+function registerHandlebarHelpers() {
+    Handlebars.registerHelper('concat', function(str1, str2) {
+        return [str1, str2].join('');
+    });
+}
+
```
Comment (Aboisier) [inline ScoutSuite/output/data/inc-scoutsuite/scoutsuite.js]: There is already a bunch of helpers registered, and I'm pretty sure there is one for concatenation... why did you add `registerHandlebarHelpers`? 

Lemme find that helper

Comment (Remi05) [reply ScoutSuite/output/data/inc-scoutsuite/scoutsuite.js]: I didn't notice the existing ones but the current `concat` joins the string with a dot (ie. `concat str1 str2` returns `str1.str2` instead of `str1str2`). I'll move mine to the `helpers.js` file and rename it.

File: ScoutSuite/providers/gcp/resources/gce/instances.py (hunk 2 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/resources/gce/instances.py]: Good idead! I wasn't a fan of having an `async` parsing method.

File: ScoutSuite/providers/gcp/provider.py (hunk 5 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/provider.py]: Forgot to uncomment it, good catch!

File: ScoutSuite/providers/gcp/resources/gce/instances.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    _children = [ 
+        (InstanceDisks, 'disks')
+    ]
+
+    def __init__(self, gcp_facade, project_id, zone):
+        self.gcp_facade = gcp_facade
+        self.project_id = project_id
+        self.zone = zone
+
+    async def fetch_all(self):
```
[the remark this answers is not in the document]
Comment (Remi05) [reply ScoutSuite/providers/gcp/resources/gce/instances.py]: Indeed 😐

[the remark this answers is not in the document]
Comment (Remi05) [reply ScoutSuite/providers/gcp/resources/gce/instances.py]: Though, since InstanceDisks is a bit different (the disks are already in the Instance, they are not fetched), it's simpler to call `fetch_all()` for each of the disks.

File: ScoutSuite/output/data/html/partials/gcp/left_menu_for_gcp_region.html (hunk 6 above)
Comment (Remi05) [reply ScoutSuite/output/data/html/partials/gcp/left_menu_for_gcp_region.html]: Merged and took care of it!
